src/java.base/share/classes/java/lang/invoke/VarHandles.java

```diff
@@ -106,66 +107,66 @@ else if (type == double.class) {
             }
         }
         else {
-            // TODO This is not lazy on first invocation
-            // and might cause some circular initialization issues
-
-            // Replace with something similar to direct method handles
-            // where a barrier is used then elided after use
-
-            if (UNSAFE.shouldBeInitialized(refc))
-                UNSAFE.ensureClassInitialized(refc);
+            if (UNSAFE.shouldBeInitialized(refc)) {
+                return new LazyStaticVarHandle(refc, f, isWriteAllowedOnFinalFields, false);
```

Comment:
This should probably call `maybeAdapt` on the result:
```suggestion
                return maybeAdapt(new LazyStaticVarHandle(refc, f, isWriteAllowedOnFinalFields, false));
```